### PR TITLE
feat(parser): Allow primitive data types to be parsed using TypeAdapter

### PR DESCRIPTION
**Pull request by sthulb**
**Issue number:** #2734

## Summary

This PR allows customers to use union types as per #2734. 

### Changes

> Please provide a summary of what's being changed

This change adopts the `TypeAdapter` from Pydantic v2 to allow the parser and base envelope to validate between the potentially different types.

### User experience

> Please share what the user experience looks like before and after this change

```python
from aws_lambda_powertools.utilities.parser import event_parser
from pydantic import BaseModel, Field
from typing import Annotated, Any, Literal, Union


class SuccessfulCallback(BaseModel):
    status: Literal["succeeded"]
    name: str
    breed: Literal["Husky", "Labrador"]


class FailedCallback(BaseModel):
    status: Literal["failed"]
    error: str


DogCallback = Annotated[
    Union[SuccessfulCallback, FailedCallback], Field(discriminator="status")
]

@event_parser(model=DogCallback)
def lambda_handler(event: DogCallback, _: Any) -> str:
    if isinstance(event, FailedCallback): # alternatively `if event.status == "failed"` (if your type checker is on top of things)
        return f"Uh oh. Had a problem: {event.error}"

```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [X] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [X] I have performed a self-review of this change
* [X] Changes have been tested
* [x] Changes are documented
* [X] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by leandrodamascena** [comment]
Before I start reviewing this PR, I'm curious as to why CI wasn't run here. Any thoughts @sthulb?

**Comment by sthulb** [comment]
I'd like to say I know why, but I don't 😂

**Comment by leandrodamascena** [comment]
Hey @sthulb, I was reviewing this PR and exploring more use cases where `TypeAdapter` can be beneficial. I believe that if we go in this direction, we can make the `Parser` feature much more flexible without introducing a significant breaking change, I really like this.

When it comes to Pydantic v2, there are no significant performance changes when using `model_validate` or `model_validate_json` and TypeAdapter with `validate_json` and `validate_python`. Instead, I noticed that when using a `dataclass` with `TypeAdapter`, the performance is better than using a Pydantic model.

In my point of view I think we can move forward with this solution and fix some small things in this PR. I'll schedule a discussion with you and @heitorlessa to talk about this.

Thanks

### Things missing in this PR

- [x] Update PR body with use case and example
- [x] Fix tests
- [x] Update messages when raising `InvalidModelTypeError` error - we now support many more types than just `BaseModel`
- [x] Update documentation - We need to improve as a whole in another PR
- [ ] Update docstrings

File: aws_lambda_powertools/utilities/parser/parser.py
Comment (sthulb) [inline aws_lambda_powertools/utilities/parser/parser.py]: I'm not particularly happy with this exception handling

**Comment by sthulb** [comment]
> Hey @sthulb, I was reviewing this PR and exploring more use cases where `TypeAdapter` can be beneficial. I believe that if we go in this direction, we can make the `Parser` feature much more flexible without introducing a significant breaking change, I really like this.
> 
> When it comes to Pydantic v2, there are no significant performance changes when using `model_validate` or `model_validate_json` and TypeAdapter with `validate_json` and `validate_python`. Instead, I noticed that when using a `dataclass` with `TypeAdapter`, the performance is better than using a Pydantic model.
> 
> In my point of view I think we can move forward with this solution and fix some small things in this PR. I'll schedule a discussion with you and @heitorlessa to talk about this.
> 
> Thanks
> 
> ### Things missing in this PR
> * [x]  Update PR body with use case and example
> * [x]  Fix tests
> * [x]  Update messages when raising `InvalidModelTypeError` error - we now support many more types than just `BaseModel`
> * [x]  Update documentation - We need to improve as a whole in another PR
> * [ ]  Update docstrings

I'm not sure the docstrings need updating, there's no material change to the signature of the functions

**Comment by leandrodamascena** [comment]
Hey Simon, as we discussed, we need to change the function signature to accept types compatibles with`TypeAdapter`.

**Comment by leandrodamascena** [comment]
Starting the review now.

File: aws_lambda_powertools/utilities/parser/envelopes/base.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/envelopes/base.py]: Hi @heitorlessa and @sthulb, always generating a model schema using TypeAdapter could impact performance here is because we are  But I was testing and there is no performance impact doing this. The test below consistently took 1-2 ms to run with `TypeAdapter` or with `APIGatewayProxyEventModel.model_validate(event)`. So, in my opinion we are good to go here.

```python
import json
from pydantic import TypeAdapter, BaseModel

from datetime import datetime
from typing import Any, Dict, List, Optional, Type, Union, Literal

from pydantic import BaseModel, model_validator
from pydantic.networks import IPvAnyNetwork

class ApiGatewayUserCertValidity(BaseModel):
        notBefore: str
        notAfter: str
    
    
class ApiGatewayUserCert(BaseModel):
    clientCertPem: str
    subjectDN: str
    issuerDN: str
    serialNumber: str
    validity: ApiGatewayUserCertValidity


class APIGatewayEventIdentity(BaseModel):
    accessKey: Optional[str] = None
    accountId: Optional[str] = None
    apiKey: Optional[str] = None
    apiKeyId: Optional[str] = None
    caller: Optional[str] = None
    cognitoAuthenticationProvider: Optional[str] = None
    cognitoAuthenticationType: Optional[str] = None
    cognitoIdentityId: Optional[str] = None
    cognitoIdentityPoolId: Optional[str] = None
    principalOrgId: Optional[str] = None
    # see #1562, temp workaround until API Gateway fixes it the Test button payload
    # removing it will not be considered a regression in the future
    sourceIp: Union[IPvAnyNetwork, Literal["test-invoke-source-ip"]]
    user: Optional[str] = None
    userAgent: Optional[str] = None
    userArn: Optional[str] = None
    clientCert: Optional[ApiGatewayUserCert] = None


class APIGatewayEventAuthorizer(BaseModel):
    claims: Optional[Dict[str, Any]] = None
    scopes: Optional[List[str]] = None


class APIGatewayEventRequestContext(BaseModel):
    accountId: str
    apiId: str
    authorizer: Optional[APIGatewayEventAuthorizer] = None
    stage: str
    protocol: str
    identity: APIGatewayEventIdentity
    requestId: str
    requestTime: str
    requestTimeEpoch: datetime
    resourceId: Optional[str] = None
    resourcePath: str
    domainName: Optional[str] = None
    domainPrefix: Optional[str] = None
    extendedRequestId: Optional[str] = None
    httpMethod: Literal["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
    path: str
    connectedAt: Optional[datetime] = None
    connectionId: Optional[str] = None
    eventType: Optional[Literal["CONNECT", "MESSAGE", "DISCONNECT"]] = None
    messageDirection: Optional[str] = None
    messageId: Optional[str] = None
    routeKey: Optional[str] = None
    operationName: Optional[str] = None

    @model_validator(mode="before")
    def check_message_id(cls, values):
        message_id, event_type = values.get("messageId"), values.get("eventType")
        if message_id is not None and event_type != "MESSAGE":
            raise ValueError("messageId is available only when the `eventType` is `MESSAGE`")
        return values


class APIGatewayProxyEventModel(BaseModel):
    version: Optional[str] = None
    resource: str
    path: str
    httpMethod: Literal["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
    headers: Dict[str, str]
    multiValueHeaders: Dict[str, List[str]]
    queryStringParameters: Optional[Dict[str, str]] = None
    multiValueQueryStringParameters: Optional[Dict[str, List[str]]] = None
    requestContext: APIGatewayEventRequestContext
    pathParameters: Optional[Dict[str, str]] = None
    stageVariables: Optional[Dict[str, str]] = None
    isBase64Encoded: bool
    body: Optional[Union[str, Type[BaseModel]]] = None

def lambda_handler(event, context):

    #
    tp = TypeAdapter(APIGatewayProxyEventModel)
    valid = tp.validate_python(event)
    #valid = APIGatewayProxyEventModel.model_validate(event)
    print(valid.path)
    
    return {
        'statusCode': 200,
        'body': json.dumps('Hello from Lambda!')
    }
```

**OUTPUT**
```sh
Function Logs
START RequestId: 5430468a-f68c-4074-a960-f14d8ab7731d Version: $LATEST
/my/path
END RequestId: 5430468a-f68c-4074-a960-f14d8ab7731d
REPORT RequestId: 5430468a-f68c-4074-a960-f14d8ab7731d	Duration: 1.28 ms	Billed Duration: 2 ms	Memory Size: 128 MB	Max Memory Used: 59 MB
```

File: aws_lambda_powertools/utilities/parser/envelopes/base.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/envelopes/base.py:20]: Please review the docstring. The `model` is not `Type[Model]` at this time.

File: aws_lambda_powertools/utilities/parser/parser.py
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/parser.py:121]: Please review the docstring. The `model` is not `Type[Model]` at this time.

Add example with dataclass, for example.

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]:
```suggestion
Powertools for AWS Lambda (Python) supports Pydantic v2.
```

File: docs/utilities/parser.md
Comment (leandrodamascena) [inline docs/utilities/parser.md]: I think we can do better here and add a link to the TypeAdapter documentation.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hi @sthulb! I'm super excited to launch this support in v3, I think we're opening up a lot of possibilities for our customers to extend the use of the parser feature with another type of data than a Pydantic model.

I did the first review, can you check please? Thank you very much for working hard here.

**Comment by leandrodamascena** [comment]
Hello @heitorlessa and @sthulb! I conducted some performance tests with `TypeAdapter` + `Union` tags and converting the event into a `TypeAdapter` on every call. I can now say that I was mistaken about my previous point regarding performance.

In cases of simple models, that is, when it's just a model with attributes and maybe even using another model, that's okay, we don't have performance problems. But when it comes to `Union` tags, we experience a significant loss of performance, which leads me to believe that it really makes sense to create some kind of ring buffer for the already converted `TypeAdapter`.

Check the results below.

#### event.json

```json
{
  "name": "Dog",
  "breed": "Husky",
  "status": "partial_success",
  "error": "partial success"
}
```

#### Converting the `Model` into `TypeAdapter` on every call

I did about 10 invocations in 2 minutes with a 128Mb Lambda.

```python
import json
from pydantic import TypeAdapter, BaseModel

from datetime import datetime
from typing import Union, Literal, Annotated

from pydantic import BaseModel, Field

class SuccessfulCallback(BaseModel):
    status: Literal["succeeded"]
    name: str
    breed: Literal["Husky", "Labrador"]


class FailedCallback(BaseModel):
    status: Literal["failed"]
    error: str
    
class TemporaryErrorCallback(BaseModel):
    status: Literal["temporary_error"]
    error: str

class PartisalSuccessCallback(BaseModel):
    status: Literal["partial_success"]
    name: str
    breed: Literal["Husky", "Labrador"]

DogCallback = Annotated[
    Union[SuccessfulCallback, FailedCallback, PartisalSuccessCallback, TemporaryErrorCallback], Field(discriminator="status")
]
# test
def lambda_handler(event, context):

    tp = TypeAdapter(DogCallback)
    
    valid = tp.validate_python(event)
    print(valid.status)
    return {
        'statusCode': 200,
        'body': json.dumps('Hello from Lambda!')
    }
```

image

#### Using `TypeAdapter` with cache

I did about 10 invocations in 2 minutes with a 128Mb Lambda.

```python
import json
from pydantic import TypeAdapter, BaseModel

from datetime import datetime
from typing import Union, Literal, Annotated

from pydantic import BaseModel, Field


class SuccessfulCallback(BaseModel):
    status: Literal["succeeded"]
    name: str
    breed: Literal["Husky", "Labrador"]


class FailedCallback(BaseModel):
    status: Literal["failed"]
    error: str
    
class TemporaryErrorCallback(BaseModel):
    status: Literal["temporary_error"]
    error: str

class PartisalSuccessCallback(BaseModel):
    status: Literal["partial_success"]
    name: str
    breed: Literal["Husky", "Labrador"]

DogCallback = Annotated[
    Union[SuccessfulCallback, FailedCallback, PartisalSuccessCallback, TemporaryErrorCallback], Field(discriminator="status")
]

cache = {}

def lambda_handler(event, context):

    if "cached" in cache:
        tp = cache["cached"]
    else:
        tp = TypeAdapter(DogCallback)
        cache["cached"] = tp
    
    valid = tp.validate_python(event)
    print(valid.status)
    return {
        'statusCode': 200,
        'body': json.dumps('Hello from Lambda!')
    }
```

image

If you'd like, we can go through more sophisticated tests and check ColdStart, p99, and other metrics, but I don't think we need to. I can say with confidence that caching the `TypeAdapter` we will have better performance.

**Comment by heitorlessa** [comment]
yeah, that makes more sense. Huge thanks for doing the benchmark so we're triple sure -- maybe try using our `LRUDict` for that, otherwise you can create a Ring Buffer using `deque` built-in data structure in Python

File: docs/utilities/parser.md
Comment (sthulb) [reply docs/utilities/parser.md]: I don't want to add the TypeAdapter ref, seems like an implementation detail – plus their docs don't really give any useful info

**Comment by leandrodamascena** [comment]
Hi @sthulb! Thank you for addressing all the feedback! I've made a few minor changes, such as creating a dedicated function to handle the cache. Since we need to use this in two places, I believe this approach makes more sense. I've also removed the pydanticv1 compatibility, as we no longer require it.

Additionally, I've created end-to-end (e2e) tests and a performance test.

@heitorlessa we need a final review here before merge.

**Review by leandrodamascena — APPROVED** [review]

**Review by heitorlessa — APPROVED** [review]
SHIIIIPPPPP IT!!

Superb work you two!

File: docs/utilities/parser.md
Comment (heitorlessa) [inline docs/utilities/parser.md]: use `primitives` instead of `arbitrary`, as `Enum` or `SQLAlchemy Models` wouldn't work here

Comment (leandrodamascena) [reply docs/utilities/parser.md]: Done

**Review by heitorlessa — APPROVED** [review]

**Pull request merged**
